lib/Bloc/workout_bloc.dart

```diff
@@ -1,123 +1,123 @@
+  final ref = await FirebaseFirestore.instance
+      .collection('users')
+      .doc(SharedPref.getStringAsync(USER_UID))
+      .collection('workouts')
+      .where('date', isGreaterThanOrEqualTo: startDate).where('date', isLessThanOrEqualTo: endDate).get();
+
+  List<Workout> workoutList = [];
+  for (int i = 0; i < ref.docs.length; i++) {
+    Workout workout = Workout.fromJsonQuery(ref.docs[i]);
+    workout.id = ref.docs[i].id;
+    workoutList.add(workout);
+  }
+  return workoutList;
+}
+
+Stream<QuerySnapshot<Map<String, dynamic>>> getWorkoutOnDate(DateTime after, DateTime before) {
+  return FirebaseFirestore.instance
+      .collection('users')
+      .doc(SharedPref.getStringAsync(USER_UID))
+      .collection('workouts')
+      .where('date', isLessThan: after)
+      .where('date', isGreaterThan: before)
+      .snapshots();
+}
+
+Future<Workout> loadWorkout(Workout workout) async {
+  List<Exercise> exerciseList = [];
+  for (int i = 0; i < workout.exercises.length; i++) {
+    Exercise exercise = await loadExercise(workout.exercises[i]);
+    exerciseList.add(exercise);
+  }
+  workout.exercises = exerciseList;
+  return workout;
+}
+
+Future<List<Workout>> getWorkoutTemplates() async {
+  final ref = await FirebaseFirestore.instance
+      .collection('users')
+      .doc(SharedPref.getStringAsync(USER_UID))
+      .collection("workouts")
+      .where('template', isEqualTo: true)
+      .get();
+  List<Workout> workoutTemplates = [];
+
+  for (int i = 0; i < ref.docs.length; i++) {
+    final data = FirebaseFirestore.instance
+        .collection('users')
+        .doc(SharedPref.getStringAsync(USER_UID))
+        .collection('workouts')
+        .doc(ref.docs[i].id)
+        .withConverter<Workout>(
+          fromFirestore: (snapshot, _) => Workout.fromJson(snapshot.data()!),
+          toFirestore: (max, _) => max.toJson(),
+        );
+    Workout workout = await data.get().then((value) => value.data()!);
+    workout.id = data.id;
+    if (!workoutTemplates.contains(workout)) {
+      workoutTemplates.add(workout);
+    }
+  }
+  return workoutTemplates;
+}
+
+void saveWorkout(Workout workout) async {
+  List exerciseList = [];
+  for (Exercise exercise in workout.exercises) {
+    exercise.setExerciseTotals();
+    exerciseList.add(await saveExercise(exercise));
+  }
+  workout.setWorkoutTotals();
+  workout.exercises = exerciseList;
+  Map<String, Object?> jsonWorkout = workout.toJson();
+  firestoreInstance
+      .collection("users")
+      .doc(SharedPref.getStringAsync(USER_UID))
+      .collection("workouts")
+      .add(jsonWorkout)
+      .then((value) {
+  });
+
+  //databaseRef.push().set(workout.toJson());
+}
+
+void deleteWorkout(Workout workout) async {
+
+  for (Exercise exercise in workout.exercises) {
+    deleteExercise(exercise);
+  }
+  await firestoreInstance
+      .collection("users")
+      .doc(SharedPref.getStringAsync(USER_UID))
```

Comment:
Since it is used multiple times here, we can declare a top-level String variable and store the value there and use it everywhere else in this file.

Reply:
@dinurymomshad for now i think that implementation works. I think it is better then declare dedicated variable, because we use the same piece of code. I will improve this with another PR as @jorre127 mention earlier :)

Reply:
Alright.